src/Proto3/Suite/DotProto/Internal.hs

```diff
@@ -0,0 +1,138 @@
+  | otherwise
+    = Right
+    . Path
+    . fmap (T.unpack . over _head toUpper)
+    . concatMap (T.splitOn ".")
+    . T.splitOn "/"
+    . Turtle.format F.fp
+    . Turtle.dropExtension
+    $ fp
+
+fatalBadModulePath :: MonadIO m => FilePath -> String -> m a
+fatalBadModulePath (Turtle.format F.fp -> fp) (T.pack -> rsn) =
+  dieLines [Neat.text|
+    Error: failed when computing the "module path" for "${fp}": ${rsn}
+
+    Please ensure that the provided path to a .proto file is specified as
+    relative to some --includeDir path and that it has the .proto suffix.
+  |]
+
+-- | @importProto searchPaths toplevel inc@ attempts to import include-relative
+-- @inc@ after locating it somewhere in the @searchPaths@; @toplevel@ is simply
+-- the path of toplevel .proto being processed so we can report it in an error
+-- message. This function terminates the program if it cannot find the file to
+-- import or if it cannot construct a valid module path from it.
+importProto :: MonadIO m
+            => [FilePath] -> FilePath -> FilePath -> m (Either ParseError DotProto)
+importProto paths (Turtle.format F.fp -> toplevelProtoText) protoFP =
+  findProto paths protoFP >>= \case
+    Found mp fp     -> parseProtoFile mp fp
+    BadModulePath e -> fatalBadModulePath protoFP e
+    NotFound        -> dieLines [Neat.text|
+      Error: while processing include statements in "${toplevelProtoText}", failed
+      to find the imported file "${protoFPText}", after looking in the following
+      locations (controlled via the --includeDir switch(es)):
+
+      $pathsText
+    |]
+  where
+    pathsText   = T.unlines (Turtle.format ("  "%F.fp) . (</> protoFP) <$> paths)
+    protoFPText = Turtle.format F.fp protoFP
+
+data FindProtoResult
+  = Found Path FilePath
+  | NotFound
+  | BadModulePath String
+  deriving (Eq, Show)
+
+-- | Attempts to locate the first (if any) filename that exists on the given
+-- search paths, and constructs the "module path" from the given
+-- include-relative filename (2nd parameter). Terminates the program with an
+-- error if the given pathname is not relative.
+findProto :: MonadIO m => [FilePath] -> FilePath -> m FindProtoResult
+findProto searchPaths protoFP
+  | Turtle.absolute protoFP = dieLines [Neat.text|
+      Error: Absolute paths to .proto files, whether on the command line or
+      in include directives, are not currently permitted; rather, all .proto
+      filenames must be relative to the current directory, or relative to some
+      search path specified via --includeDir.
+
+      This is because we currently use the include-relative name to decide
+      the structure of the Haskell module tree that we emit during code
+      generation.
+      |]
+  | otherwise = case toModulePath protoFP of
+      Left e -> pure (BadModulePath e)
+      Right mp -> do
+        mfp <- flip Turtle.fold FL.head $ do
+          fp <- (</> protoFP) <$> Turtle.select searchPaths
+          True <- Turtle.testfile fp
```

Comment:
Minor suggestion. You could move the `</> protoFP` to here:

```
True <- Turtle.testfile (fp </> protoFP)
```

Which simplifies the `Turtle.select`, to me. This is non-critical feedback, though.